### PR TITLE
mgr/orch: take a single placement argument

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

File: src/pybind/mgr/orchestrator/_interface.py [hunk 1]
```diff
@@ -1306,7 +1306,25 @@ def from_strings(cls, strings):
...
+                strings = arg.split(';')
+            elif ',' in arg and '[' not in arg:
+                # FIXME: this isn't quite right.  we want to avoid breaking
+                # a list of mons with addrvecs... so we're basically allowing
+                # , most of the time, except when addrvecs are used.  maybe
+                # ok?
+                strings = arg.split(',')
+            else:
+                strings = [arg]
+        elif isinstance(arg, list):
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator/_interface.py]: weird, this conflicts with the annotation of this method signature where `arg` can only be `None` or `str`.

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -1306,7 +1306,25 @@ def from_strings(cls, strings):
         >>> PlacementSpec.from_strings(None)
         PlacementSpec()
```
Comment (jschmid1) [inline src/pybind/mgr/orchestrator/_interface.py]: Would you mind extending the docstring with examples?

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -760,7 +760,7 @@ def _apply_mgr(self, placement=None):
 
     @_cli_write_command(
         'orch apply mon',
-        'name=placement,type=CephString,n=N,req=false',
+        'name=placement,type=CephString,req=false',
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator/module.py]: this breaks the test of `tasks.mgr.test_orchestrator_cli.TestOrchestratorCli`:

```
2020-03-04T09:58:53.303 INFO:teuthology.orchestra.run.smithi203:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph --log-early log 'Ended test tasks.mgr.test_orchestrator_cli.TestOrchestratorCli.test_mon_update'
2020-03-04T09:58:54.465 INFO:tasks.cephfs_test_runner:
2020-03-04T09:58:54.466 INFO:tasks.cephfs_test_runner:======================================================================
2020-03-04T09:58:54.466 INFO:tasks.cephfs_test_runner:ERROR: test_mon_update (tasks.mgr.test_orchestrator_cli.TestOrchestratorCli)
2020-03-04T09:58:54.466 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-04T09:58:54.466 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-03-04-1337/qa/tasks/mgr/test_orchestrator_cli.py", line 160, in test_mon_update
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:    self._orch_cmd("apply", "mon", "3", "host1:1.2.3.0/24", "host2:1.2.3.0/24", "host3:10.0.0.0/8")
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-03-04-1337/qa/tasks/mgr/test_orchestrator_cli.py", line 21, in _orch_cmd
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:    return self._cmd("orch", *args)
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-03-04-1337/qa/tasks/mgr/test_orchestrator_cli.py", line 18, in _cmd
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:    return self.mgr_cluster.mon_manager.raw_cluster_cmd(module, *args)
2020-03-04T09:58:54.467 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-03-04-1337/qa/tasks/ceph_manager.py", line 1359, in raw_cluster_cmd
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:    stdout=BytesIO(),
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 198, in run
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 433, in run
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:    r.wait()
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 158, in wait
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:    self._raise_for_status()
2020-03-04T09:58:54.468 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 180, in _raise_for_status
2020-03-04T09:58:54.469 INFO:tasks.cephfs_test_runner:    node=self.hostname, label=self.label
2020-03-04T09:58:54.469 INFO:tasks.cephfs_test_runner:CommandFailedError: Command failed on smithi203 with status 22: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph --log-early orch apply mon 3 host1:1.2.3.0/24 host2:1.2.3.0/24 host3:10.0.0.0/8'
```

see `/a//kchai-2020-03-04_09:24:42-rados-wip-kefu-testing-2020-03-04-1337-distro-basic-smithi/4824403`

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

File: src/pybind/mgr/orchestrator/module.py (hunk 3 above)
Comment (liewegas) [reply src/pybind/mgr/orchestrator/module.py]: fixed

File: src/pybind/mgr/orchestrator/_interface.py (hunk 1 above)
Comment (liewegas) [reply src/pybind/mgr/orchestrator/_interface.py]: yeah, i'm waiting on #33476 and also coding defensively... but probably best to let the other PR merge and then adjust all callers to pass a string.

Comment (liewegas) [reply src/pybind/mgr/orchestrator/_interface.py]: rebased and cleaned up

**Comment by mgfritch** [comment]
mypy run-test: commands[0] | mypy --config-file=../../mypy.ini cephadm/module.py cephadm/_utils.py mgr_module.py dashboard/module.py mgr_util.py orchestrator/__init__.py progress/module.py rook/module.py osd_support/module.py test_orchestrator/module.py volumes/__init__.py
cephadm/module.py:1882: error: Type signature has too few arguments
Found 1 error in 1 file (checked 11 source files)

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
green: http://pulpito.ceph.com/sage-2020-03-07_12:43:42-rados:cephadm-wip-sage-testing-2020-03-06-2310-distro-basic-smithi/

**Comment by liewegas** [comment]
@tchaikov want to weigh in here?

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-03-07_07:27:51-rados-wip-kefu-testing-2020-03-07-1143-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-07_16:58:24-rados-wip-kefu-testing-2020-03-07-2053-distro-basic-smithi/

File: src/pybind/mgr/orchestrator/_interface.py [hunk 4]
```diff
@@ -1312,28 +1312,44 @@ def validate(self):
...
         PlacementSpec(label=[HostSpec(hostname='host1', network='', name=''), HostSpec(hostname='host2', network='', name='')])
         You can also prefix the hosts with a count as follows:
-        >>> PlacementSpec.from_strings('2 host1 host2'.split())
+        >>> PlacementSpec.from_string('2 host1 host2')
         PlacementSpec(label=[HostSpec(hostname='host1', network='', name=''), HostSpec(hostname='host2', network='', name='')], count=2)
         You can spefify labels using `label:<label>`
-        >>> PlacementSpec.from_strings('label:mon'.split())
+        >>> PlacementSpec.from_string('label:mon')
         PlacementSpec(label='label:mon')
         Labels als support a count:
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator/_interface.py]: [off-topic] s/als/also/

**Comment by liewegas** [comment]
> i don't quite understand if we should tolerant the case where the specified count is not consistent with the number of following hosts.. or it does not matter? shall we error out by throwing some exception to inform user that the input is invalid?
> 
> but given that this changeset does not change this behavior. LGTM..

Currently if the count is < the host list, it will choose from among them.  And if it's > it will choose however many additional hosts it needs from the rest of the cluster.  But it's on my todo list to sit down and try to document this clearly...

**Pull request merged**